File: iOS/ai/Data/AnalysisProviders.swift
import Foundation

// MARK: - Analysis Providers

extension AnalysisRepository {

    // MARK: - Dispatch by Format

    func analyze(
        service: AppService,
        apiKey: String,
        prompt: String,
        model: String,
        params: AgentParameters? = nil,
        baseUrl: String? = nil
    ) async -> AnalysisResponse {
        switch service.apiFormat {
        case .anthropic:
            return await analyzeWithClaude(service: service, apiKey: apiKey, prompt: prompt, model: model, params: params)
        case .google:
            return await analyzeWithGemini(service: service, apiKey: apiKey, prompt: prompt, model: model, params: params)
        case .openaiCompatible:
            return await analyzeWithOpenAiCompatible(service: service, apiKey: apiKey, prompt: prompt, model: model, params: params, baseUrl: baseUrl ?? service.baseUrl)
        }
    }

    // MARK: - OpenAI Compatible

    func analyzeWithOpenAiCompatible(
        service: AppService,
        apiKey: String,
        prompt: String,
        model: String,
        params: AgentParameters? = nil,
        baseUrl: String = ""
    ) async -> AnalysisResponse {
        let effectiveBaseUrl = baseUrl.isEmpty ? service.baseUrl : baseUrl

        // Check Responses API
        if service.usesResponsesApi(model: model) {
            return await analyzeWithResponsesApi(service: service, apiKey: apiKey, prompt: prompt, model: model, params: params)
        }

        // Build messages
        var messages: [OpenAiMessage] = []
        if let sp = params?.systemPrompt, !sp.isEmpty {
            messages.append(OpenAiMessage(role: "system", content: sp))
        }
        messages.append(OpenAiMessage(role: "user", content: prompt))

        // Build seed fields
        let seedValue = params?.seed
        let isMistralSeed = service.seedFieldName == "random_seed"

        let request = OpenAiRequest(
            model: model,
            messages: messages,
            max_tokens: params?.maxTokens,
            temperature: params?.temperature,
            top_p: params?.topP,
            top_k: params?.topK,
            frequency_penalty: params?.frequencyPenalty,
            presence_penalty: params?.presencePenalty,
            stop: params?.stopSequences?.isEmpty == false ? params?.stopSequences : nil,
            seed: isMistralSeed ? nil : seedValue,
            random_seed: isMistralSeed ? seedValue : nil,
            response_format: params?.responseFormatJson == true ? OpenAiResponseFormat(type: "json_object") : nil,
            return_citations: service.supportsCitations ? params?.returnCitations : nil,
            search_recency_filter: service.supportsSearchRecency ? params?.searchRecency : nil,
            search: params?.searchEnabled == true ? true : nil
        )

        let url = ApiClient.chatUrl(for: service, customBaseUrl: effectiveBaseUrl)
        let headers = ApiClient.openAiHeaders(apiKey: apiKey)

        do {
            let body = try ApiClient.encode(request)
            let (response, statusCode, respHeaders): (OpenAiResponse, Int, [String: String]) =
                try await ApiClient.shared.sendRequest(url: url, headers: headers, body: body, responseType: OpenAiResponse.self)

            // Extract content with multi-strategy
            let content = response.choices?.first?.message.content
                ?? response.choices?.first?.message.reasoning_content
                ?? response.choices?.compactMap({ $0.message.content }).first
                ?? response.choices?.compactMap({ $0.message.reasoning_content }).first

            guard let content = content else {
                return AnalysisResponse(service: service, error: "No response content (choices: \(response.choices?.count ?? 0))",
                                        httpStatusCode: statusCode)
            }

            let tokenUsage = response.usage.map {
                TokenUsage(
                    inputTokens: $0.prompt_tokens ?? $0.input_tokens ?? 0,
                    outputTokens: $0.completion_tokens ?? $0.output_tokens ?? 0,
                    apiCost: PricingCache.extractApiCost(openAiUsage: $0, service: service.extractApiCost ? service : nil)
                )
            }

            return AnalysisResponse(
                service: service,
                analysis: content,
                tokenUsage: tokenUsage,
                citations: response.citations,
                searchResults: response.search_results,
                relatedQuestions: response.related_questions,
                rawUsageJson: response.usage != nil ? Self.formatUsageJson(response.usage) : nil,
                httpHeaders: Self.formatHeaders(respHeaders),
                httpStatusCode: statusCode
            )
        } catch let error as ApiError {
            return AnalysisResponse(service: service, error: error.localizedDescription)
        } catch {
            return AnalysisResponse(service: service, error: "Error: \(error.localizedDescription)")
        }
    }

    // MARK: - OpenAI Responses API

    private func analyzeWithResponsesApi(
        service: AppService,
        apiKey: String,
        prompt: String,
        model: String,
        params: AgentParameters? = nil
    ) async -> AnalysisResponse {
        let request = OpenAiResponsesRequest(
            model: model,
            input: .text(prompt),
            instructions: params?.systemPrompt?.isEmpty == false ? params?.systemPrompt : nil
        )

        let normalizedBase = ApiClient.normalizeBaseUrl(service.baseUrl)
        let url = "\(normalizedBase)v1/responses"
        let headers = ApiClient.openAiHeaders(apiKey: apiKey)

        do {
            let body = try ApiClient.encode(request)
            let (data, statusCode, respHeaders) = try await ApiClient.shared.sendRawRequest(url: url, headers: headers, body: body)

            guard (200...299).contains(statusCode) else {
                let errorBody = String(data: data, encoding: .utf8) ?? "Unknown error"
                return AnalysisResponse(service: service, error: "API error: \(statusCode) - \(errorBody)", httpStatusCode: statusCode)
            }

            let response = try ApiClient.decode(OpenAiResponsesApiResponse.self, from: data)

            // Multi-strategy content extraction
            let firstOutput = response.output?.first
            let firstContent = firstOutput?.content
            let outputText: String? = firstContent?.first(where: { $0.type == "output_text" })?.text
            let textType: String? = firstContent?.first(where: { $0.type == "text" })?.text
            let anyText: String? = firstContent?.compactMap({ $0.text }).first
            let messageOutput = response.output?.first(where: { $0.type == "message" })
            let messageText: String? = messageOutput?.content?.compactMap({ $0.text }).first
            let flatText: String? = response.output?.flatMap({ $0.content ?? [] }).compactMap({ $0.text }).first
            let content: String? = outputText ?? textType ?? anyText ?? messageText ?? flatText

            guard let content = content else {
                return AnalysisResponse(service: service, error: "No response content from Responses API", httpStatusCode: statusCode)
            }

            let tokenUsage = response.usage.map {
                TokenUsage(
                    inputTokens: $0.prompt_tokens ?? $0.input_tokens ?? 0,
                    outputTokens: $0.completion_tokens ?? $0.output_tokens ?? 0,
                    apiCost: PricingCache.extractApiCost(openAiUsage: $0, service: service.extractApiCost ? service : nil)
                )
            }

            return AnalysisResponse(
                service: service,
                analysis: content,
                tokenUsage: tokenUsage,
                rawUsageJson: response.usage != nil ? Self.formatUsageJson(response.usage) : nil,
                httpHeaders: Self.formatHeaders(respHeaders),
                httpStatusCode: statusCode
            )
        } catch {
            return AnalysisResponse(service: service, error: "Error: \(error.localizedDescription)")
        }
    }

    // MARK: - Anthropic / Claude

    func analyzeWithClaude(
        service: AppService,
        apiKey: String,
        prompt: String,
        model: String,
        params: AgentParameters? = nil
    ) async -> AnalysisResponse {
        let request = ClaudeRequest(
            model: model,
            max_tokens: params?.maxTokens ?? 4096,
            messages: [ClaudeMessage(role: "user", content: prompt)],
            temperature: params?.temperature,
            top_p: params?.topP,
            top_k: params?.topK,
            system: params?.systemPrompt?.isEmpty == false ? params?.systemPrompt : nil,
            stop_sequences: params?.stopSequences?.isEmpty == false ? params?.stopSequences : nil,
            frequency_penalty: params?.frequencyPenalty,
            presence_penalty: params?.presencePenalty,
            seed: params?.seed,
            search: params?.searchEnabled == true ? true : nil
        )

        let normalizedBase = ApiClient.normalizeBaseUrl(service.baseUrl)
        let url = "\(normalizedBase)v1/messages"
        let headers = ApiClient.anthropicHeaders(apiKey: apiKey)

        do {
            let body = try ApiClient.encode(request)
            let (response, statusCode, respHeaders): (ClaudeResponse, Int, [String: String]) =
                try await ApiClient.shared.sendRequest(url: url, headers: headers, body: body, responseType: ClaudeResponse.self)

            if let error = response.error {
                return AnalysisResponse(service: service, error: "Claude error: \(error.message ?? error.type ?? "Unknown")",
                                        httpStatusCode: statusCode)
            }

            let content = response.content?.first(where: { $0.type == "text" })?.text
                ?? response.content?.compactMap({ $0.text }).first

            guard let content = content else {
                return AnalysisResponse(service: service, error: "No response content from Claude", httpStatusCode: statusCode)
            }

            let tokenUsage = response.usage.map {
                TokenUsage(
                    inputTokens: $0.input_tokens ?? 0,
                    outputTokens: $0.output_tokens ?? 0,
                    apiCost: PricingCache.extractApiCost(claudeUsage: $0)
                )
            }

            return AnalysisResponse(
                service: service,
                analysis: content,
                tokenUsage: tokenUsage,
                rawUsageJson: response.usage != nil ? Self.formatUsageJson(response.usage) : nil,
                httpHeaders: Self.formatHeaders(respHeaders),
                httpStatusCode: statusCode
            )
        } catch let error as ApiError {
            return AnalysisResponse(service: service, error: error.localizedDescription)
        } catch {
            return AnalysisResponse(service: service, error: "Error: \(error.localizedDescription)")
        }
    }

    // MARK: - Google Gemini

    func analyzeWithGemini(
        service: AppService,
        apiKey: String,
        prompt: String,
        model: String,
        params: AgentParameters? = nil
    ) async -> AnalysisResponse {
        let request = GeminiRequest(
            contents: [GeminiContent(parts: [GeminiPart(text: prompt)])],
            generationConfig: GeminiGenerationConfig(
                temperature: params?.temperature,
                topP: params?.topP,
                topK: params?.topK,
                maxOutputTokens: params?.maxTokens,
                stopSequences: params?.stopSequences?.isEmpty == false ? params?.stopSequences : nil,
                frequencyPenalty: params?.frequencyPenalty,
                presencePenalty: params?.presencePenalty,
                seed: params?.seed,
                search: params?.searchEnabled == true ? true : nil
            ),
            systemInstruction: params?.systemPrompt?.isEmpty == false ? GeminiContent(parts: [GeminiPart(text: params!.systemPrompt!)]) : nil
        )

        let url = ApiClient.geminiGenerateUrl(baseUrl: service.baseUrl, model: model, apiKey: apiKey)
        let headers = ApiClient.geminiHeaders()

        do {
            let body = try ApiClient.encode(request)
            let (response, statusCode, respHeaders): (GeminiResponse, Int, [String: String]) =
                try await ApiClient.shared.sendRequest(url: url, headers: headers, body: body, responseType: GeminiResponse.self)

            if let error = response.error {
                return AnalysisResponse(service: service, error: "Gemini error: \(error.message ?? "Unknown") (\(error.status ?? ""))",
                                        httpStatusCode: statusCode)
            }

            let content = response.candidates?.first?.content?.parts.first?.text
                ?? response.candidates?.first?.content?.parts.compactMap({ $0.text }).first
                ?? response.candidates?.flatMap({ $0.content?.parts ?? [] }).compactMap({ $0.text }).first

            guard let content = content else {
                return AnalysisResponse(service: service, error: "No response content from Gemini", httpStatusCode: statusCode)
            }

            let tokenUsage = response.usageMetadata.map {
                TokenUsage(
                    inputTokens: $0.promptTokenCount ?? 0,
                    outputTokens: $0.candidatesTokenCount ?? 0,
                    apiCost: PricingCache.extractApiCost(geminiUsage: $0)
                )
            }

            return AnalysisResponse(
                service: service,
                analysis: content,
                tokenUsage: tokenUsage,
                rawUsageJson: response.usageMetadata != nil ? Self.formatUsageJson(response.usageMetadata) : nil,
                httpHeaders: Self.formatHeaders(respHeaders),
                httpStatusCode: statusCode
            )
        } catch let error as ApiError {
            return AnalysisResponse(service: service, error: error.localizedDescription)
        } catch {
            return AnalysisResponse(service: service, error: "Error: \(error.localizedDescription)")
        }
    }
}

File: android/ai/src/main/java/com/ai/data/AnalysisProviders.kt
package com.ai.data


/**
 * Unified analysis method for OpenAI-format providers that support the Responses API.
 * For providers with apiFormat == OPENAI_COMPATIBLE, the analyzeWithOpenAiCompatible method
 * handles most cases. This handles the OpenAI-specific Responses API for newer models.
 */
internal suspend fun AnalysisRepository.analyzeWithChatGptResponsesApi(service: AppService, apiKey: String, prompt: String, model: String, params: AgentParameters? = null): AnalysisResponse {
    val api = ApiFactory.createOpenAiApiWithBaseUrl(service.baseUrl)
    val request = OpenAiResponsesRequest(
        model = model,
        input = prompt,
        instructions = params?.systemPrompt?.takeIf { it.isNotBlank() }
    )
    val response = api.createResponse(
        authorization = "Bearer $apiKey",
        request = request
    )

    val headers = formatHeaders(response.headers())
    val statusCode = response.code()
    return if (response.isSuccessful) {
        val body = response.body()
        val content = body?.output?.let { outputs ->
            outputs.firstOrNull()?.content?.firstOrNull { it.type == "output_text" }?.text
                ?: outputs.firstOrNull()?.content?.firstOrNull { it.type == "text" }?.text
                ?: outputs.firstOrNull()?.content?.firstNotNullOfOrNull { it.text }
                ?: outputs.firstOrNull { it.type == "message" }?.content?.firstNotNullOfOrNull { it.text }
                ?: outputs.flatMap { it.content ?: emptyList() }.firstNotNullOfOrNull { it.text }
        }
        val rawUsageJson = formatUsageJson(body?.usage)
        val usage = body?.usage?.let {
            TokenUsage(
                inputTokens = it.input_tokens ?: it.prompt_tokens ?: 0,
                outputTokens = it.output_tokens ?: it.completion_tokens ?: 0,
                apiCost = extractApiCost(it)
            )
        }
        if (content != null) {
            AnalysisResponse(service, content, null, usage, rawUsageJson = rawUsageJson, httpHeaders = headers, httpStatusCode = statusCode)
        } else {
            val errorMsg = body?.error?.message ?: "No response content (output: ${body?.output})"
            AnalysisResponse(service, null, errorMsg, httpHeaders = headers, httpStatusCode = statusCode)
        }
    } else {
        val errorBody = try { response.errorBody()?.string() } catch (_: Exception) { null }
        val errorDetail = if (errorBody != null) "API error: ${response.code()} ${response.message()} - $errorBody" else "API error: ${response.code()} ${response.message()}"
        AnalysisResponse(service, null, errorDetail, httpHeaders = headers, httpStatusCode = statusCode)
    }
}

internal suspend fun AnalysisRepository.analyzeWithClaude(service: AppService, apiKey: String, prompt: String, model: String, params: AgentParameters? = null): AnalysisResponse {
    val claudeApi = ApiFactory.createClaudeApiWithBaseUrl(service.baseUrl)
    val request = ClaudeRequest(
        model = model,
        messages = listOf(ClaudeMessage(role = "user", content = prompt)),
        max_tokens = params?.maxTokens ?: 4096,
        temperature = params?.temperature,
        top_p = params?.topP,
        top_k = params?.topK,
        system = params?.systemPrompt?.takeIf { it.isNotBlank() },
        stop_sequences = params?.stopSequences?.takeIf { it.isNotEmpty() },
        frequency_penalty = params?.frequencyPenalty,
        presence_penalty = params?.presencePenalty,
        seed = params?.seed,
        search = if (params?.searchEnabled == true) true else null
    )
    val response = claudeApi.createMessage(apiKey = apiKey, request = request)

    val headers = formatHeaders(response.headers())
    val statusCode = response.code()
    return if (response.isSuccessful) {
        val body = response.body()
        val content = body?.content?.let { contentBlocks ->
            contentBlocks.firstOrNull { it.type == "text" }?.text
                ?: contentBlocks.firstNotNullOfOrNull { it.text }
        }
        val rawUsageJson = formatUsageJson(body?.usage)
        val usage = body?.usage?.let {
            TokenUsage(
                inputTokens = it.input_tokens ?: 0,
                outputTokens = it.output_tokens ?: 0,
                apiCost = extractApiCost(it)
            )
        }
        if (content != null) {
            AnalysisResponse(service, content, null, usage, rawUsageJson = rawUsageJson, httpHeaders = headers, httpStatusCode = statusCode)
        } else {
            val errorMsg = body?.error?.message ?: "No response content (blocks: ${body?.content?.size ?: 0})"
            AnalysisResponse(service, null, errorMsg, httpHeaders = headers, httpStatusCode = statusCode)
        }
    } else {
        val errorBody = try { response.errorBody()?.string() } catch (_: Exception) { null }
        val errorMsg = if (errorBody != null) "API error: ${response.code()} ${response.message()} - $errorBody" else "API error: ${response.code()} ${response.message()}"
        AnalysisResponse(service, null, errorMsg, httpHeaders = headers, httpStatusCode = statusCode)
    }
}

internal suspend fun AnalysisRepository.analyzeWithGemini(service: AppService, apiKey: String, prompt: String, model: String, params: AgentParameters? = null): AnalysisResponse {
    // Build generation config with all parameters
    val generationConfig = if (params != null) {
        GeminiGenerationConfig(
            temperature = params.temperature,
            topP = params.topP,
            topK = params.topK,
            maxOutputTokens = params.maxTokens,
            stopSequences = params.stopSequences?.takeIf { it.isNotEmpty() },
            frequencyPenalty = params.frequencyPenalty,
            presencePenalty = params.presencePenalty,
            seed = params.seed,
            search = if (params.searchEnabled) true else null
        )
    } else null

    // System instruction for Gemini
    val systemInstruction = params?.systemPrompt?.takeIf { it.isNotBlank() }?.let {
        GeminiContent(parts = listOf(GeminiPart(text = it)))
    }

    val request = GeminiRequest(
        contents = listOf(
            GeminiContent(parts = listOf(GeminiPart(text = prompt)))
        ),
        generationConfig = generationConfig,
        systemInstruction = systemInstruction
    )

    val geminiApi = ApiFactory.createGeminiApiWithBaseUrl(service.baseUrl)
    val response = geminiApi.generateContent(model = model, apiKey = apiKey, request = request)

    android.util.Log.d("GeminiAPI", "Response code: ${response.code()}, message: ${response.message()}")

    val headers = formatHeaders(response.headers())
    val statusCode = response.code()
    return if (response.isSuccessful) {
        val body = response.body()
        // Try multiple parsing strategies for content extraction
        val content = body?.candidates?.let { candidates ->
            // Strategy 1: Get text from first candidate's first part
            candidates.firstOrNull()?.content?.parts?.firstOrNull()?.text
            // Strategy 2: Get any non-null text from first candidate's parts
                ?: candidates.firstOrNull()?.content?.parts?.firstNotNullOfOrNull { it.text }
            // Strategy 3: Try any candidate with non-null content
                ?: candidates.flatMap { it.content?.parts ?: emptyList() }.firstNotNullOfOrNull { it.text }
        }
        val rawUsageJson = formatUsageJson(body?.usageMetadata)
        val usage = body?.usageMetadata?.let {
            TokenUsage(
                inputTokens = it.promptTokenCount ?: 0,
                outputTokens = it.candidatesTokenCount ?: 0,
                apiCost = extractApiCost(it)
            )
        }
        if (content != null) {
            AnalysisResponse(service, content, null, usage, rawUsageJson = rawUsageJson, httpHeaders = headers, httpStatusCode = statusCode)
        } else {
            val errorMsg = body?.error?.message ?: "No response content (candidates: ${body?.candidates?.size ?: 0})"
            AnalysisResponse(service, null, errorMsg, httpHeaders = headers, httpStatusCode = statusCode)
        }
    } else {
        val errorBody = try { response.errorBody()?.string() } catch (_: Exception) { null }
        android.util.Log.e("GeminiAPI", "Error body: $errorBody")
        AnalysisResponse(service, null, "API error: ${response.code()} ${response.message()} - $errorBody", httpHeaders = headers, httpStatusCode = statusCode)
    }
}

/**
 * Unified analysis method for all OpenAI-compatible providers.
 * Handles provider-specific request fields and response extraction.
 * Also handles the Responses API for OpenAI newer models (gpt-5.x, o3, o4).
 */
internal suspend fun AnalysisRepository.analyzeWithOpenAiCompatible(
    service: AppService,
    apiKey: String,
    prompt: String,
    model: String,
    params: AgentParameters? = null,
    baseUrl: String = service.baseUrl
): AnalysisResponse {
    // Check if this model uses the Responses API (OpenAI gpt-5.x, o3, o4)
    if (usesResponsesApi(service, model)) {
        return analyzeWithChatGptResponsesApi(service, apiKey, prompt, model, params)
    }

    val api = ApiFactory.createOpenAiCompatibleApi(baseUrl)
    val normalizedBase = if (baseUrl.endsWith("/")) baseUrl else "$baseUrl/"
    val chatUrl = normalizedBase + service.chatPath

    // Build messages with optional system prompt
    val messages = buildList {
        params?.systemPrompt?.let { systemPrompt ->
            if (systemPrompt.isNotBlank()) {
                add(OpenAiMessage(role = "system", content = systemPrompt))
            }
        }
        add(OpenAiMessage(role = "user", content = prompt))
    }

    val request = OpenAiRequest(
        model = model,
        messages = messages,
        max_tokens = params?.maxTokens,
        temperature = params?.temperature,
        top_p = params?.topP,
        top_k = params?.topK,
        frequency_penalty = params?.frequencyPenalty,
        presence_penalty = params?.presencePenalty,
        stop = params?.stopSequences?.takeIf { it.isNotEmpty() },
        seed = if (service.seedFieldName == "seed") params?.seed else null,
        random_seed = if (service.seedFieldName == "random_seed") params?.seed else null,
        response_format = if (params?.responseFormatJson == true) OpenAiResponseFormat(type = "json_object") else null,
        return_citations = if (service.supportsCitations) params?.returnCitations else null,
        search_recency_filter = if (service.supportsSearchRecency) params?.searchRecency else null,
        search = if (params?.searchEnabled == true) true else null
    )

    val response = api.createChatCompletion(
        url = chatUrl,
        authorization = "Bearer $apiKey",
        request = request
    )

    val headers = formatHeaders(response.headers())
    val statusCode = response.code()
    return if (response.isSuccessful) {
        val body = response.body()
        // Try multiple parsing strategies for content extraction
            val content = body?.choices?.let { choices ->
                choices.firstOrNull()?.message?.content
                    ?: choices.firstOrNull()?.message?.reasoning_content
                    ?: choices.firstNotNullOfOrNull { it.message.content }
                    ?: choices.firstNotNullOfOrNull { it.message.reasoning_content }
            }
        val citations = body?.citations
        val searchResults = body?.search_results
        val relatedQuestions = body?.related_questions
        val rawUsageJson = formatUsageJson(body?.usage)
        val usage = body?.usage?.let {
            TokenUsage(
                inputTokens = it.prompt_tokens ?: it.input_tokens ?: 0,
                outputTokens = it.completion_tokens ?: it.output_tokens ?: 0,
                apiCost = extractApiCost(it, if (service.extractApiCost) service else null)
            )
        }
        if (!content.isNullOrBlank()) {
            AnalysisResponse(service, content, null, usage, citations = citations, searchResults = searchResults, relatedQuestions = relatedQuestions, rawUsageJson = rawUsageJson, httpHeaders = headers, httpStatusCode = statusCode)
        } else {
            val errorMsg = body?.error?.message ?: "No response content (choices: ${body?.choices?.size ?: 0})"
            AnalysisResponse(service, null, errorMsg, httpHeaders = headers, httpStatusCode = statusCode)
        }
    } else {
        val errorBody = try { response.errorBody()?.string() } catch (_: Exception) { null }
        val errorDetail = if (errorBody != null) "API error: ${response.code()} ${response.message()} - $errorBody" else "API error: ${response.code()} ${response.message()}"
        AnalysisResponse(service, null, errorDetail, httpHeaders = headers, httpStatusCode = statusCode)
    }
}
